File: 프로그래머스/3/43238. 입국심사/입국심사.java
class Solution {

    public long solution(int n, int[] times) {
        long answer = 0L;
        long left = 0L;
        long right = n * (long) 1e9;

        while (left <= right) {
            long mid = (left + right) >>> 1;
            long count = simulate(n, times, mid);

            if (count < n) {
                left = mid + 1;
            } else {
                answer = mid;
                right = mid - 1;
            }
        }

        return answer;
    }

    long simulate(int n, int[] times, long givenTime) {
        long count = 0L;

        for (int time : times) {
            count += givenTime / time;
        }

        return count;
    }

}

File: 프로그래머스/3/43238. 입국심사/입국심사.kt
class Solution {
    
    private val MAX = 1e18.toLong(); 
    
    fun solution(n: Int, times: IntArray): Long {
        var left = 0L;
        var right = MAX;
        var answer = right;
        
        while (left <= right) {
            val mid = (left + right) ushr 1;
            val cnt = simulate(mid, times);
            
            if (cnt < n) {
                left = mid + 1;
            } else {
                right = mid - 1;
                answer = mid;
            }
        }
        
        return answer
    }
    
    private fun simulate(k: Long, times: IntArray): Long {
        var cnt = 0L;
        
        for (time: Int in times) {
            cnt += k / time;
        }
        
        return cnt;
    } 
}